pkg/storage/http/lister.go

```diff
@@ -0,0 +1,113 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"golang.org/x/net/html"
+)
+
+// List lists all versions of a module
+func (s *ModuleStore) List(ctx context.Context, module string) ([]string, error) {
+	const op errors.Op = "http.List"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+
+	baseURL := s.moduleRoot(module)
+
+	req, _ := http.NewRequest(http.MethodGet, baseURL, nil)
+	req.SetBasicAuth(s.username, s.password)
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		kind := errors.KindUnexpected
+		if resp.StatusCode == http.StatusNotFound {
+			// This is a unique case because this is actually okay! It means that there
+			// are no versions here.
+			return []string{}, nil
+		}
+		return nil, errors.E(op, kind, errors.M(module))
+	}
+
+	// Find all the links to ".mod" files within the current module
+	//
+	// Ok, so admittedly this scheme for listing versions is a little harebrained.
+	// But that said it _is_ pretty standard for directory indexes to be formatted
+	// very closely to what we expect here. I'm not settled on this but it does
+	// work surprisingly well.
+	mods, err := collectLinks(resp.Body, func(ref string) bool {
+		bu, _ := url.Parse(baseURL)
+		bu.Path = path.Join(bu.Path, ref)
+		absolute := bu.String()
+		return strings.HasPrefix(absolute, baseURL) && strings.HasSuffix(absolute, ".mod")
+	})
+
+	// convert mod file references to versions
+	versions := make([]string, len(mods))
+	for i, m := range mods {
+		versions[i] = strings.TrimSuffix(path.Base(m), ".mod")
+	}
+
+	return versions, nil
+}
+
+// collectLinks traverses the contents of r, which is assumed to be HTML, and gathers
+// all <a> tag references that pass filter.
+func collectLinks(r io.Reader, filter func(string) bool) ([]string, error) {
+
+	t := html.NewTokenizer(r)
+
+	var links []string
+	var err error
+
+floop:
```

Comment:
Personally I feel like this label makes reading and reasoning about this code confusing. Do you think there's a way to remove it so that the code is more understandable and not way longer?

Reply:
Good suggestion. I addressed this with an early return, which did simplify things slightly.